### ISSUE TITLE
About the result

### ISSUE THREAD

**Issue by austingg** [yusuketomoto/chainer-fast-neuralstyle#1]
First, awesome work, seem the first re-implementation of the paper.

Currently,  the result seem a little worse than the vanilla neural style or the result in "perceptual loss" paper. maybe the low tv weights result in the noisy and grid like results.  Are all the hyperparameter the same as paper's ?

**Comment by yusuketomoto** [comment]
Thanks.

Yes, I know my result is a little block noisy. 
The model which this repo contains was not trained with the same hyper parameter as paper's.
I didn't know the exact value of lambda_feat and lambda_style when I trained the model, because I couldn't found them in the paper.
But I found them [here](https://www.reddit.com/r/MachineLearning/comments/4cj1jj/160308155_perceptual_losses_for_realtime_style/), and gonna update repo!

It seemed lambda_style which I used was too large.
And I think that uniform lambda_style weight might not be good, because I found that gram_matrix losses are quite different from each layer.

**Comment by austingg** [comment]
Yes, there are so many hyper-parameters began from the seminal work "neural algorithm for artistic stylization". Currently most re-implementation follows uniform lambda_style weights. I have done lots for experiments based justin's implementation. Seems your current result only learned the low-level style, e.g. color, no higher-level such as strokes. 
And in the reddit blog , justin offer a supplementary material for more details. may be helpful for u.

**Comment by dionysio** [comment]
@yusuketomoto any news? What weights have you tried? 

I have just started training with lambda_style=5 and lambda_feat=1

**Comment by graphific** [comment]
it might be that your convolution kernel size is 4? whereas the original VGG uses 3, which should give more precise gfx.

**Comment by mionyu** [comment]
Thx for sharing codes. I trained with default parameters in this repo on mscoco 2014 dataset, tried several style images including starrynight, results are all noisy and weird. Could anyone reproduce the default  starrynight.model in this repo?

**Comment by valentinvieriu** [comment]
@mionyu  I'm getting the same here. I've trained on several pictures and getting unusable results. I'm getting a strange frame affect on all the models , no matter what the source is.
I really wanted to use this in an art project next weekend, but unfortunately the results are not acceptable at all. See the examples. Any help is welcomed. I've used the standard command line without altering any parameters.
![image](https://cloud.githubusercontent.com/assets/478866/16373819/0e62c0ce-3c54-11e6-99e6-07e04b0c17ee.png)
![image](https://cloud.githubusercontent.com/assets/478866/16373853/32de4784-3c54-11e6-81d8-e7c8513a29bc.png)

**Comment by yusuketomoto** [comment]
Try batchsize 1, you might get better result.
I can't have confidence, but I'm suspicious of batch_matmul or style loss computation with batch.

![mod_15000](https://cloud.githubusercontent.com/assets/3320497/16378678/21af57a6-3ca8-11e6-8379-f7e97ba691d8.png)

![muse_s_15500](https://cloud.githubusercontent.com/assets/3320497/16378685/2e791580-3ca8-11e6-8ed3-a32beddc5134.png)

**Comment by warmspringwinds** [comment]
@yusuketomoto Hello.

Thank you for a nice implementation.

I have one question:
I keep getting strange spots with random values when I apply the algorithm.
Exactly like on the last picture with cat that you have sent on the right side right near the border.
Do you know if there is a way to avoid it?

Thank you.

**Comment by valentinvieriu** [comment]
Thank you @yusuketomoto . The batchsize 1 did help. The image looks more nice and appealing now.
![image](https://cloud.githubusercontent.com/assets/478866/16404402/f8d8d0b2-3cfe-11e6-802e-207cff076d4c.png)

The border is still there. If you have any idea how to remove that, would be appreciated. Do you think it has anything to do with the training data?

**Comment by nikolajspetersen** [comment]
@yusuketomoto, Thanks for a great implementation here - really enjoy playing with it. Since the features of the style image are being replicated in a pretty granular way, when the input image is larger than style img size, I wondered if you have tried to train with with larger resolution than the 256, 256 that the code resizes it to do. 

I guess training will take proportionally longer time then. Just curious if you have experience with doing that? Is there a way to alter the model to create larger features in the output image once its trained?

**Comment by 6o6o** [comment]
@nikolajspetersen, I'm curious about this too. Furthermore, since the resolution is always 256x256, regardless of the aspect ratio, it tends to distort shapes in training data. Correct me if i'm wrong, or this is irrelevant.

Currently I'm training a model on 512 px style image and center crops of training data of the same size. Will post back comparison images once it's ready.

**Comment by 6o6o** [comment]
Our assumptions were correct. Size does matter. The result is much more appealing. Here's my picture processed in style of [Kandinsky painting](http://www.wikiart.org/en/wassily-kandinsky/transverse-line-1923). Both models were trained with batchsize 1 for 2 epochs.

First uses the default size of 256.
Second uses 512 px style and square centered crops of training data, rescaled with bicubic interpolation so that the smallest side is 512. Training took 44 hours on GTX 970.

The downsides are that cropping looses some of the detail, wile upscaling blurs the image slightly. I guess you can get rid of cropping and squeeze whatever image it is into square dimensions, don't know if it will turn out in feature distortion though.

![ya_kandinsky](https://cloud.githubusercontent.com/assets/13799992/16719315/a890b4b6-4731-11e6-8237-862de8fd1883.png)

![ya_crop512](https://cloud.githubusercontent.com/assets/13799992/16719358/06badc4c-4732-11e6-90ce-5e37e41281ff.png)

**Comment by codevui** [comment]
@6o6o I'd try trained with my gt1080 - 8GB with size 512, but it inform that out of memory. 
Do you have anyway to reduce memory ?

**Comment by codevui** [comment]
@6o6o nice work!
I'm using gtx 1080 - 8GB ram and try to train with size 512 but get out of memory at "feature_hat = vgg(y)". Can you tell me how to reduce memory size ?

**Comment by 6o6o** [comment]
@codevui thanks. Check if you have everything right. Basically, just replace all instances of 256 with the desired value and optionally set resampling filter to 2. It shouldn't be using more than 3600 MB. Actually, 8GB should allow you to go all the way up to 720px

**Comment by bucktoothsir** [comment]
@6o6o nice work. it really helps me a lot.
by the way, could you please tell me the exact value of lambda_feat and lambda_style you used?

**Comment by codevui** [comment]
@6o6o the problem is cudnn. After i make cudnn correct for chainer, everything is ok!
Thank you!

**Comment by 6o6o** [comment]
@bucktoothsir, glad to help. I haven't touched those values, as I'm not sure about them. Experimenting with trial and error takes too much time. I noticed though, in the paper possible range for `lambda_tv` is between 1e-4 and 1e-6, while here it's 10e-4. Don't know if it's a typo, or am I missing something. You can try adjusting `lambda_tv` and `lambda_style` and post back if anything interesting comes up

**Comment by gafr** [comment]
@6o6o  would you mind sharing your model?

I am also currently training a new one (256), due to the time intensive model creation, we should consider opening a central repo with some models. What do you think?

**Comment by 6o6o** [comment]
@gafr, sure. I'm all for it. Would be great if we could collect some good models for everyone to use. Should I just create a new repo? Or where do I put them?

**Comment by gafr** [comment]
@6o6o  just invited you to a repository, give me 10 minutes for a README and a structure

**Comment by yusuketomoto** [comment]
As I found subtracting mean images before image transformation matrix is not effective, I'd like to change codes.
However this change will break backward compatibility of models.
If you create a repository, please note that.

**Comment by gafr** [comment]
@yusuketomoto would you give me the permission to collect/share your current model in a repository, or should I just link it to your repository. I will note down the parameters and/or version

**Comment by yusuketomoto** [comment]
@gafr I updated the codes, models, readme text. Sorry for the inconvenience!

**Comment by gafr** [comment]
Some collected models are now available at https://github.com/gafr/chainer-fast-neuralstyle-models

**Comment by logic1988** [comment]
@6o6o The second result is great!  I am training the model but the result was terrible. Can you tell me some details?

**Comment by 6o6o** [comment]
1. Reducing the dataset to 10k significantly deteriorates the quality. Definitely not recommended.
2. Scale the images so that the smallest side is of desired value, preserving the aspect ratio, then crop. This is optional. I read that VGG models were trained like this [here](https://gist.github.com/ksimonyan/fd8800eeb36e276cd6f9#description), so I thought the same technique could be applied to styles. Changes should be done to train.py only.
3. All parameters at default. You can try increasing `--lambda_tv` to smth like 10e-3. Lowering it as proposed in paper tends to produce more artifacts.

If you wish I can fork the project and integrate the changes. I don't think PR is a good idea since the enhanced settings take considerably longer to train and may not be suitable for everyone.

**Comment by rogozin70** [comment]
@6o6o Write to me at my email - rogozin_70@inbox.ru
 I would like to cooperate with you.

**Comment by rayset** [comment]
Anyone imolemented a video version? It should be simple considering the speed of this alghoritm

**Comment by logic1988** [comment]
@6o6o Thank you very much! Please fork the project then I can learn from it. Looking forward to more communication.

**Comment by 6o6o** [comment]
@rayset I've done some experiments. The result is quite steady, preserving consistency between frames for moving objects / transitions, except for some jitter in the backgrounds, areas with subtle color change, compression artifacts, etc. See [here](https://youtu.be/15c_hppAkk8) and [here](https://youtu.be/_QxIh_mN1-4)

**Comment by rayset** [comment]
Do you mind sharing your code for the frame by frame conversion? Even by pm if it's not clean, I'm not really good with scripts and I would love a basis to start from . 

Thanks in advance

**Comment by austingg** [comment]
@yusuketomoto  have u ever fix the batchsize( >=2) problem?  mini-batch traininig will be faster and stable

**Comment by yusuketomoto** [comment]
@austingg No, I'm afraid not. As reconstruction without style losses work fine, there would be some bugs in style loss computation with minibatch.

**Comment by ttoinou** [comment]
I still have some glitchs, here is a close-up, does anyone know why and how to avoid them ?

![harmonia1_a_00862_eaufeuille3b_256_1_40000 _glitchs](https://cloud.githubusercontent.com/assets/12092576/17223472/02433aa0-54fd-11e6-963b-a1f7cd623791.jpg)

**Comment by rayset** [comment]
I'm also fighting  with those dots, I'd love to get riferimento of them

**Comment by austingg** [comment]
@ttoinou what is your training params and epochs?

**Comment by ttoinou** [comment]
@austingg : I used default params, 1,5 epochs (1 epoch at 80k images, and 40k images from the second epoch)

**Comment by rayset** [comment]
same but with 2 epochs. old version tho.

**Comment by austingg** [comment]
@rayset  I have trained a model with defautl params and 2 epochs with imagent val (50k), seem quite smooth

**Comment by austingg** [comment]
@yusuketomoto @6o6o @rayset  
how to remove the border artifact? It come out in the first epoch, and seem does get better in the second epoch

**Comment by ttoinou** [comment]
@austingg Would you have a link to your image set ?
If you want to get rid of the border artifact and can't fix the algorithm, you could always feed the neural network with a bigger image (black or white borders, or even repetitions of image's original pixels)

**Comment by austingg** [comment]
@ttoinou I use ILSVRC2012 val dataset for training. 

I know what you said, first extend the image, and generating image and then crop.  
It's weird, cause sometime there are no border artifact. 

Maybe the problem is 4x4 conv filter or tv, but i am not sure. I will work on this problem.

**Comment by 6o6o** [comment]
@austingg do you train on cropped images or fullsize? I find fullsize training to decrease border and pixelation a bit, but produces artifacts. See [crop](https://github.com/gafr/chainer-fast-neuralstyle-models/blob/master/images/kandinsky_e2_crop512.jpg?raw=true) vs. [fullsize](https://github.com/gafr/chainer-fast-neuralstyle-models/blob/master/images/kandinsky_e2_full512.jpg?raw=true)

**Comment by austingg** [comment]
@6o6o  I use fullsize (just stretch the image to 256x256 or 512x512)

**Comment by rayset** [comment]
Eh eh did you found out about ILSVRC2012 from deepart.io?
I was just messing around with it.
There must be something wrong in most implementation(dots and border,a almost every single one on GitHub has them).

Do you still get the border on the most recent version? What about dots?

**Comment by 6o6o** [comment]
The project was updated. Several changes were integrated, including the ones I was talking about in my [above comment](#issuecomment-231638046), mainly adjustable training image size and retaining aspect ratio of dataset images by cropping, but done slightly differently.

Whether to crop or not to crop remains an open question. I made a short [slideshow](https://drive.google.com/folderview?id=0B5o-5frMSje-alIyWU9PV0ZLT3M&usp=sharing) for comparison. I think cropping is generally more appropriate, because in most cases it gets rid of useless areas on photos, e.g. backgrounds, irrelevant details or insignificant features. Only in some rare cases I would prefer not to crop, like img10 and img20, where we loose a person's face, which might be important. I find cropped result to be superior, except for more pronounced pixel dithering pattern and a frame border around the picture, which are not as evident in a full version, but some artifacts occur there. But again, it's only one style, others may have different effects.

I updated my fork as well, some of the differences are
* toggle cropping with `--fullsize` option (default is to crop)
* different cropping implementation. I just use built-in function [`ImageOps.fit`](http://pillow.readthedocs.io/en/3.1.x/reference/ImageOps.html#PIL.ImageOps.fit), which produces images in the slideshow above 
* bicubic and Lanczos resampling when scaling dataset and input style images provides sharper shrinking

See readme in my [fork](https://github.com/6o6o/chainer-fast-neuralstyle) for more info. @logic1988 you might be interested in it.
@rayset I also included a quick and dirty video processing bash script `genvid.sh`. It extracts frames, transforms them one at a time and then composes back together. Tested with YouTube .mp4 videos. The speed could be optimized by excluding model loading on every iteration, but it's ok for now.

**Comment by rayset** [comment]
@6o6o awesome fork, I noticed that in the update they added all the feaures you discussed. 
may I ask you an exampe of the usage of genvid? I replaced the first rows like this:

INPUT=primo.mp4
OUTPUT=video.mp4
MODEL=models/composition.model


and it works till the estraction. in row 16 I get a bad substitution error (I guess when you try (?) to truncate the first 7 characters from f. what am I doing wrong?

**Comment by 6o6o** [comment]
hmm, weird, I've not had that error, maybe try executing
`chmod 755 genvid.sh` and then `bash genvid.sh`?

**Comment by rayset** [comment]
you were right! works perfect now.

I'm reading about frame consistency: https://github.com/manuelruder/artistic-videos do you think it would be hard to implement this procedure? the video looks pretty nice even now, but I wonder if it can be improved

**Comment by 6o6o** [comment]
Ah, that algorithm is superb, the result are very [impressive](https://youtu.be/vbbTb5xZSig). I haven't tried it, but from what I heard it's
* much harder to implement since you need to couple video rendering with image generation
* much slower, since it's based on original neural-style, even though it shouldn't redraw overlapping frame areas, but still...

This one is much faster and easier to use, therefor better suited for general public. As for optimization / reducing jitter in some areas, it maybe be helpful to process frames with some sort of median filter before transformation to smooth out noise and subtle details but keep the edges rough. I'll do some more experiments will let you know if anything interesting comes up.

**Comment by ttoinou** [comment]
@6o6o : great ! Thanks for the updates. I also forked your project :) !
Look this post : https://github.com/yusuketomoto/chainer-fast-neuralstyle/issues/25 .

**Comment by cryptexis** [comment]
Hi everyone, 
this implementation uses 4 size kernel and I am pretty sure a lot of people(including me) tried to use kernel size 3. But whoever tried it, knows that deconvolution layer does not handle sizes correctly. 
So, this pull request deals with it :) Not sure if it's gonna be accepted by the main repo :)
[ddd8643360282a074d9fcbdcedbac470713132d3](https://github.com/pfnet/chainer/pull/1467/commits/ddd8643360282a074d9fcbdcedbac470713132d3)

Good luck :)

**Comment by rayset** [comment]
Does the kernel size of 3 really remove the border artifact and the dots?

**Comment by cryptexis** [comment]
emmm not sure yet :) it's in the process. 
Also, do you still get borders when using batch_size 1?

**Comment by rayset** [comment]
Yep ,the default value is 1

**Comment by rogozin70** [comment]
Really, no one else can create a quality model for "yusuketomoto/chainer-fast-neuralstyle"
photo processing? A total of 6 models, but they look good - kandinsky, edtaonisl.
If use a similar pattern?

**Comment by ttoinou** [comment]
What would be a quality model for photo processing ?!

What is the benefit of using kernel size 3 ?

**Comment by cryptexis** [comment]
@ttoinou several reasons I see there.
1) being consistent with the original paper, as the paper states that the kernel size should be 3 on 2nd and 3rd convolution layers. Also on 1st and 2nd deconvolution layers.
2) Less parameters to train
3) Also, I think usually kernel size and other parameters are cross-validated. So it's not a coincidence that it was chosen to be 3.

You can experiment as much as you want :) I just gave the ability to do so.

**Comment by austingg** [comment]
@yusuketomoto  @cryptexis  @rayset  
I think that the border artifact is produced by the total variance module, which is implemented by convolution function with zero padding, so the borders have stronger gradients.

**Comment by 6o6o** [comment]
@cryptexis thanks. Do I also have to change the `pad` parameter in the neighbouring layers? Haven't found anything in the papers about it.

**Comment by mstrazds** [comment]
@yusuketomoto great implementation, really enjoying working with it. I thought I would share a render of some test footage using @6o6o 's Kandinsky model. @rayset definitely keen on looking for a way to implement https://github.com/manuelruder/artistic-videos in some way. Here is the 1080p output: https://youtu.be/sKOa3tgHQho (youtube does make a bit of a meal of it) and a comparison with the source: https://youtu.be/ovMKXzkPMsw

Example still:
![example-still](https://cloud.githubusercontent.com/assets/756378/17518316/d07a728a-5e8a-11e6-8ff1-06a5c5ccd517.jpg)

I am training some models now, and will be sure to share them when done.

**Comment by randomrandom** [comment]
hey @mstrazds, how did you blend the frames together? did you use optical flow or just simple blending?

**Comment by DylanAlloy** [comment]
@mstrazds I notice that the artifacts and dots are much harder to find in the video implementation. Why do you think that is?

**Comment by mstrazds** [comment]
@DylanAlloy When the video was encoded on youtube it reduced the quality significantly. As a part of this the 'dots' or pixels seem to be either removed to the high compression or harder to perceive. The video was constructed by rendering out stills from generate.py and then stitching them together using ffmpeg using x264 for compression. There were no tricks for merging frames or anything of the like.

**Comment by rupeshs** [comment]
Check out http://neuralstyler.com/
Turn Your Videos/Photos/GIF into Art

https://github.com/rupeshs/neuralstyler
